### PR TITLE
Feature/admin panel

### PR DESCRIPTION
**Pull request by umaimaz123**

### CODE REVIEW

File: app/javascript/packs/application.js [hunk 1]
```diff
@@ -15,5 +15,7 @@ require("channels")
 //
 // const images = require.context('../images', true)
 // const imagePath = (name) => images(name, true)
+//= require bootstrap
+//= require jquery
 import "bootstrap"
 import "../stylesheets/application"
```
Comment (Sohair63) [inline app/javascript/packs/application.js:21]: what is the difference  between require and import?

File: app/models/user.rb [hunk 2]
```diff
@@ -21,6 +21,10 @@ def self.find_for_database_authentication(warden_condition)
     ).first
   end
 
+  def self.filter(search)
+    where('username LIKE ? OR email LIKE ? OR full_name LIKE ?', "%#{search}%", "%#{search}%", "%#{search}%")
+  end
```
Comment (Sohair63) [inline app/models/user.rb:26]: how about downcase and search, so you will be able to search case insensitive

File: app/views/admin/users/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,66 @@
+<div class="jumbotron">
+  <h1 class="display-6"><%= params[:action].titleize %>Client</h1>
+  <hr class="my-4">
+  
+  <% if user.errors.any? %>
+  <h2>Errors:</h2>
+    <ul>
```
Comment (Sohair63) [inline app/views/admin/users/_form.html.erb:7]: lint

File: app/views/admin/users/_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    <% user.errors.full_messages.each do |error| %>
+      <li><%= error %></li>
+    <% end %>
+    </ul>
+  <% end %>
+  
+  <%= form_for ([:admin, @user]) do |f| %> 
+  <div class="mb-3 row">
+    <%= f.label :username, class: "col-sm-2 col-form-label" %></br>
```
Comment (Sohair63) [inline app/views/admin/users/_form.html.erb:16]: lint

File: app/views/admin/users/_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  </div>
+
+  <div class="mb-3 row">
+    <%= f.label :full_name, class: "col-sm-2 col-form-label" %></br>
+    <div class="col-sm-10">
+      <%= f.text_field :full_name, required: true, class:"form-control" %>
+    </div>
+  </div>
+
+  <%# if params[:action] == 'edit' %>
```
Comment (Sohair63) [inline app/views/admin/users/_form.html.erb:36]: remove

File: app/views/admin/users/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    <%= f.label :password_confirmation, class: "col-sm-2 col-form-label" %></br>
+    <div class="col-sm-10">
+      <%= f.text_field :password_confirmation, required: true, class:"form-control" %>
+    </div>
+  </div>
+  <div class="text-secondary d-flex justify-content-center">
+    Hint: The password should be at least eight characters long.</br>
+    To make it stronger, use upper and lower case letters, numbers, and symbols like ! " ? $ % ^ & ).
+ </div>
+  <%# end %>
```
Comment (Sohair63) [inline app/views/admin/users/_form.html.erb:54]: remove

File: app/views/admin/users/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+</div>
+<% end %>
+<div class="d-flex flex-row-reverse">
+  <div class="p-1">
+    <%= link_to 'New', new_admin_user_path, class:"btn btn-dark" %> 
+  </div>
+  <div class="p-1">
+    <%= link_to 'Back', admin_dashboard_index_path, class:"btn btn-dark" %> 
+  </div>
+
```
Comment (Sohair63) [inline app/views/admin/users/index.html.erb:17]: remove line

File: app/views/admin/users/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,45 @@
+<h1>All Users</h1>
+<%= form_tag(admin_users_path, method: "get") do %>
+<div class="input-group">
```
Comment (Sohair63) [inline app/views/admin/users/index.html.erb:3]: lint

File: app/views/admin/users/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    <tr>
+      <th scope="col">User name</th>
+      <th scope="col">Email</th>
+      <th scope="col">Full name</th>
+      <th scope="col">Actions</th>
+    </tr>
+  </thead>
+  <tbody>
+    <% @users&.each do |user| %>
+    <tr>
```
Comment (Sohair63) [inline app/views/admin/users/index.html.erb:31]: you do not need & here if this is collection coming from controller

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -10,10 +10,16 @@
     <%= stylesheet_pack_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
   </head>
 
+<script src="https://unpkg.com/@popperjs/core@2"></script>
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
+<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>
```
Comment (Sohair63) [inline app/views/layouts/application.html.erb:15]: you are adding this in assets and also including here, this should not be included here.

File: config/routes.rb [hunk 11]
```diff
@@ -9,6 +9,11 @@
 
   namespace :admin do
     resources :dashboard
+    resources :users do
+      member do
+        get 'delete_modal'
```
Comment (Sohair63) [inline config/routes.rb:14]: is there any better way, so we do not include this modal?

**Review by Sohair63 — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 2 above)
Comment (umaimaz123) [reply app/models/user.rb:26]: ok

File: app/javascript/packs/application.js (hunk 1 above)
Comment (umaimaz123) [reply app/javascript/packs/application.js:21]: require will scan the node modules. thanks for pointing out. I will remove the import bootstrap one.

File: app/views/layouts/application.html.erb (hunk 10 above)
Comment (umaimaz123) [reply app/views/layouts/application.html.erb:15]: ok

File: config/routes.rb (hunk 11 above)
Comment (umaimaz123) [reply config/routes.rb:14]: I found this method to show modal after much search. should i use sweet alerts as i discussed it with miss momna and she suggest me this or should i use anything else?

File: app/models/user.rb [hunk 12]
```diff
@@ -21,6 +21,11 @@ def self.find_for_database_authentication(warden_condition)
     ).first
   end
 
+  def self.filter(search)
+      param = search&.downcase 
+      where('username LIKE ? OR email LIKE ? OR full_name LIKE ?', "%#{param}%", "%#{param}%", "%#{param}%")
+  end
```
Comment (Sohair63) [inline app/models/user.rb:27]:
```ruby
where('lowecase(username) LIKE :term OR lowecase(email) LIKE :term OR lowecase(full_name) LIKE :term', term: param)
```

**Review by Sohair63 — CHANGES_REQUESTED** [review]

**Pull request merged**